File: frontend/src/api/categories.ts
import { api } from './client'
import type { Category, CategoryIn, CategoryUpdate } from '@/types'

export const categoriesApi = {
  list: () => api.get<Category[]>('/categories').then((r) => r.data),

  get: (id: string) => api.get<Category>(`/categories/${id}`).then((r) => r.data),

  create: (data: CategoryIn) => api.post<Category>('/categories', data).then((r) => r.data),

  update: (id: string, data: CategoryUpdate) =>
    api.put<Category>(`/categories/${id}`, data).then((r) => r.data),

  delete: (id: string) => api.delete(`/categories/${id}`),
}

File: backend/app/routers/categories.py
import uuid
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.user import User
from app.schemas.category import CategoryIn, CategoryOut, CategoryUpdate
from app.services import category_service
from app.dependencies import get_current_user

router = APIRouter(prefix="/categories", tags=["categories"])


@router.post("", response_model=CategoryOut, status_code=status.HTTP_201_CREATED)
def create(
    data: CategoryIn,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.create(db, data, current_user.id)


@router.get("", response_model=list[CategoryOut])
def list_all(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.list_all(db, current_user.id)


@router.get("/{category_id}", response_model=CategoryOut)
def get_one(
    category_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.get_one(db, category_id, current_user.id)


@router.put("/{category_id}", response_model=CategoryOut)
def update(
    category_id: uuid.UUID,
    data: CategoryUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return category_service.update(db, category_id, data, current_user.id)


@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete(
    category_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    category_service.delete(db, category_id, current_user.id)
